db/rdb.go

```diff
@@ -438,52 +439,70 @@ func (r *RDBDriver) GetByCpeURI(uri string) ([]models.CveDetail, error) {
 		}
 	}
 
-	idDetail := map[uint]models.CveDetail{}
-	for _, cpe := range filtered {
-		var cveDetailID uint
+	return filtered, nil
+}
+
+// GetIDByCpeURI Select Cve Ids by by pseudo-CPE
+func (r *RDBDriver) GetIDByCpeURI(uri string) ([]string, error) {
+	filtered, err := r.getMatchingCpe(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	cveIDs := make([]string, len(filtered))
+
+	// The cpes table is de-normalized. So the `First` is correct, don't be mislead.
+	for i, cpe := range filtered {
 		if cpe.JvnID != 0 {
 			jvn := models.Jvn{}
-			err = r.conn.Select("cve_detail_id").Where("ID = ?", cpe.JvnID).First(&jvn).Error
+			err = r.conn.Select("cve_id").Where("ID = ?", cpe.JvnID).First(&jvn).Error
 			if err != nil && err != gorm.ErrRecordNotFound {
 				return nil, err
 			}
-			cveDetailID = jvn.CveDetailID
+			cveIDs[i] = jvn.CveID
 		} else if cpe.NvdXMLID != 0 {
 			nvd := models.NvdXML{}
-			err = r.conn.Select("cve_detail_id").Where("ID = ?", cpe.NvdXMLID).First(&nvd).Error
+			err = r.conn.Select("cve_id").Where("ID = ?", cpe.NvdXMLID).First(&nvd).Error
 			if err != nil && err != gorm.ErrRecordNotFound {
 				return nil, err
 			}
-			cveDetailID = nvd.CveDetailID
+			cveIDs[i] = nvd.CveID
 		} else if cpe.NvdJSONID != 0 {
 			json := models.NvdJSON{}
-			err = r.conn.Select("cve_detail_id").Where("ID = ?", cpe.NvdJSONID).First(&json).Error
+			err = r.conn.Select("cve_id").Where("ID = ?", cpe.NvdJSONID).First(&json).Error
 			if err != nil && err != gorm.ErrRecordNotFound {
 				return nil, err
 			}
-			cveDetailID = json.CveDetailID
+			cveIDs[i] = json.CveID
 		}
 
-		if _, ok := idDetail[cveDetailID]; ok {
-			continue
+		// If we didn't find a CVE something is weird.
+		if cveIDs[i] == "" {
+			log.Infof("Missing cve_id for %s (id: %d)", cpe.URI, cpe.ID)
 		}
+	}
 
-		cveDetail := models.CveDetail{}
-		err = r.conn.Select("cve_id").Where("ID = ?", cveDetailID).First(&cveDetail).Error
-		if err != nil && err != gorm.ErrRecordNotFound {
-			return nil, err
-		}
-		cveID := cveDetail.CveID
-		detail, err := r.Get(cveID)
-		if err != nil {
-			return nil, err
-		}
-		idDetail[cveDetailID] = *detail
+	return cveIDs, nil
+}
+
+// GetByCpeURI Select Cve information from DB.
+func (r *RDBDriver) GetByCpeURI(uri string) ([]models.CveDetail, error) {
+	cveIDs, err := r.GetIDByCpeURI(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	cveDetails, err := r.GetMulti(cveIDs)
+	if err != nil {
+		return nil, err
 	}
 
-	details := []models.CveDetail{}
-	for _, d := range idDetail {
-		details = append(details, d)
+	// Convert from map to array
+	details := make([]models.CveDetail, len(cveDetails))
+	i := 0
+	for _, d := range cveDetails {
```

Comment:
```suggestion
	for i, d := range cveDetails {
	    details[i] = d
```
Is this not good enough?

Reply:
`cveDetails` comes back as a map, but the existing function signature is an array. `for i, d := range cveDetails` will enumerate by key, value, and not index, item.

Reply:
I understand. I'm sorry that I could not read properly.

Reply:
I did the same thing when I wrote it. :)